gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitConfigMonitor.java

```diff
@@ -0,0 +1,419 @@
+        ConfigurationKeys.DEFAULT_GIT_CONFIG_MONITOR_REPO_DIR);
+
+    this.configDir = ConfigUtils.getString(config, ConfigurationKeys.GIT_CONFIG_MONITOR_CONFIG_DIR,
+        ConfigurationKeys.DEFAULT_GIT_CONFIG_MONITOR_CONFIG_DIR);
+
+    this.pollingInterval = ConfigUtils.getInt(config, ConfigurationKeys.GIT_CONFIG_MONITOR_POLLING_INTERVAL,
+        ConfigurationKeys.DEFAULT_GIT_CONFIG_MONITOR_POLLING_INTERVAL);
+
+    this.configDirPath = new Path(this.repositoryDir, this.configDir);
+
+    try {
+      this.pullFileLoader = new PullFileLoader(this.configDirPath,
+          FileSystem.get(URI.create(ConfigurationKeys.LOCAL_FS_URI), new Configuration()),
+          PullFileLoader.DEFAULT_JAVA_PROPS_PULL_FILE_EXTENSIONS, PullFileLoader.DEFAULT_HOCON_PULL_FILE_EXTENSIONS);
+    } catch (IOException e) {
+      throw new RuntimeException("Could not create pull file loader", e);
+    }
+
+    try {
+      this.gitRepo = new GitRepository(repositoryUri, this.repositoryDir);
+    } catch (GitAPIException | IOException e) {
+      throw new RuntimeException("Could not open git repository", e);
+    }
+  }
+
+  /** Start the service. */
+  @Override
+  protected void startUp() throws Exception {
+    log.info("Starting the " + GitConfigMonitor.class.getSimpleName());
+    log.info("Polling git with inteval {} ", this.pollingInterval);
+
+    // Schedule the job config fetch task
+    this.scheduledExecutor.scheduleAtFixedRate(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          processGitConfigChanges();
+        } catch (GitAPIException | IOException e) {
+          log.error("Failed to process git config changes", e);
+          // next run will try again since errors could be intermittent
+        }
+      }
+    }, 0, this.pollingInterval, TimeUnit.SECONDS);
+  }
+
+  /** Stop the service. */
+  @Override
+  protected void shutDown() throws Exception {
+    this.scheduledExecutor.shutdown();
+    this.scheduledExecutor.awaitTermination(TERMINATION_TIMEOUT, TimeUnit.SECONDS);
+  }
+
+  public synchronized void setActive(boolean isActive) {
+    if (this.isActive == isActive) {
+      // No-op if already in correct state
+      return;
+    }
+
+    this.isActive = isActive;
+  }
+
+  /**
+   * Fetch the list of changes since the last refresh of the repository and apply the changes to the {@link FlowCatalog}
+   * @throws GitAPIException
+   * @throws IOException
+   */
+  @VisibleForTesting
+  void processGitConfigChanges() throws GitAPIException, IOException {
+    // if not active or if the flow catalog is not up yet then can't process config changes
+    if (!isActive || !this.flowCatalog.isRunning()) {
+      return;
+    }
+
+    List<DiffEntry> changes = this.gitRepo.getChanges();
+
+    for (DiffEntry change : changes) {
+      switch (change.getChangeType()) {
+        case ADD:
+        case MODIFY:
+          addSpec(change);
```

Comment:
MODIFY is the update case. The `FlowCatalog` handles update in the put() call like add.
```
if (specStore.exists(spec.getUri())) {
        specStore.updateSpec(spec);
        this.listeners.onUpdateSpec(spec);
      } else {
```